Update backtraces proposal to current implementation (#6)
bgamari/ghc-proposals: proposals/0000-exception-backtraces.rst

root of the hierarchy: ::

    class (Typeable e, Show e) => Exception e where
        -- | Represent the exception as 'SomeExceptionWithBacktrace'
        -- If @e@ isn't already of type 'SomeExceptionWithBacktrace' this means some kind of wrapping.
        toException   :: e -> SomeExceptionWithBacktrace

>         -- | Extract and cast the exception from it's wrapped representation

Reviewer: ```suggestion
        -- | Extract and cast the exception from its wrapped representation
```

Representing and capturing backtraces
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> GHC has at four distinct mechanisms for capturing backtraces, each with

Reviewer: ```suggestion
GHC has four distinct mechanisms for capturing backtraces, each with
```

> it's own backtrace representation:

Reviewer: ```suggestion
its own backtrace representation:
```

We propose that ``GHC.Exception`` provides a family of these functions for
the ``HasCallStack``, cost-center stack, info table provenance (IPE) and execution
stack cases. ::

    throwIOWithCallStack :: (HasCallStack, Exception e) => e -> IO a

    throwIOWithIPEStack :: Exception e => e -> IO a

    throwIOWithCostCenterStack :: Exception e => e -> IO a

    throwIOWithExecutionStack :: Exception e => e -> IO a

We also propose to apply the same ideas to ``throw``, keeping some symmetry
between ``GHC.Exception`` and ``GHC.IO``. ::

    module GHC.Exception where

    throwWithCallStack :: HasCallStack => forall (r :: RuntimeRep). forall (a :: TYPE r). forall e.
            Exception e => e -> a

    throwWithIPEStack :: forall (r :: RuntimeRep). forall (a :: TYPE r). forall e.
            Exception e => e -> a

    throwWithCostCenterStack :: forall (r :: RuntimeRep). forall (a :: TYPE r). forall e.
            Exception e => e -> a

    throwWithExecutionStack :: forall (r :: RuntimeRep). forall (a :: TYPE r). forall e.
            Exception e => e -> a

(The prototype implementation showed that these functions do not add much code.)

Pretty Printing
~~~~~~~~~~~~~~~

> The be compliant with the convention that ``Show`` instances should output

Reviewer: ```suggestion
To be compliant with the convention that ``Show`` instances should output
```

> valid Haskell code, the `Show` instance of ``Backtrace`` delegates to the

Reviewer: ```suggestion
valid Haskell code, the ``Show`` instance of ``Backtrace`` delegates to the
```

        CostCenterBacktraceMech
      | -- | use execution stack unwinding with given limit
        ExecutionStackBacktraceMech
      | -- | collect backtraces from Info Table Provenance Entries
        IPEBacktraceMech
      | -- | use 'HasCallStack'
      HasCallStackBacktraceMech
      deriving (Eq, Show)

    currentBacktraceMechanisms :: IORef [BacktraceMechanism]
    currentBacktraceMechanisms = unsafePerformIO $ newIORef []
    {-# NOINLINE currentBacktraceMechanisms #-}

    -- | Set how 'Control.Exception.throwIO', et al. collect backtraces.
    setDefaultBacktraceMechanisms :: [BacktraceMechanism] -> IO ()
    setDefaultBacktraceMechanisms = writeIORef currentBacktraceMechanisms

    -- | Returns the currently selected 'BacktraceMechanism'.
    getDefaultBacktraceMechanisms :: IO [BacktraceMechanism]
    getDefaultBacktraceMechanisms = readIORef currentBacktraceMechanisms

A ``collectBacktrace`` primitive used by ``throw`` (and likewise ``throwIO``)
simply dispatches to the appropriate backtrace collection scheme as determined
by the currently selected ``BacktraceMechanism``s: ::

    module GHC.Exception.Backtrace where

>     -- | Collect a list of 'Backtrace' via all current default 'BacktraceMechanism'.

Reviewer: ```suggestion
    -- | Collect a list of 'Backtrace's via all current default 'BacktraceMechanism's.
```

    -- See 'setDefaultBacktraceMechanisms'

>     collectBacktraces :: HasCallStack =>IO [Backtrace]

Reviewer: ```suggestion
    collectBacktraces :: HasCallStack => IO [Backtrace]
```

    collectBacktraces = do
        mechs <- getDefaultBacktraceMechanisms
        catMaybes `fmap` mapM collectBacktraces' mechs
      where
        -- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
        collectBacktraces' :: HasCallStack => BacktraceMechanism -> IO (Maybe Backtrace)
        collectBacktraces' CostCenterBacktraceMech = collectCostCenterBacktrace
        collectBacktraces' ExecutionStackBacktraceMech = collectExecutionStackBacktrace
        collectBacktraces' IPEBacktraceMech = collectIPEBacktrace
        collectBacktraces' HasCallStackBacktraceMech = collectHasCallStackBacktrace

    module GHC.Exception where

    -- | Throw an exception. Exceptions may be thrown from purely
    -- functional code, but may only be caught within the 'IO' monad.
    -- 'Backtrace' backtraces are collected according to the configured
    -- 'BacktraceMechanism's.
    --

>     -- WARNING: You may want to use 'throwIO' instead so that your pure code

Reviewer: ```suggestion
    -- WARNING: If you are in an `IO` context you may want to rather use 'throwIO' instead so that your pure code
```

cross-cutting concern. While a stateless approach would be preferred, we
believe that this compromise is a significant improvement over the status quo.

> The Haskell community will have to adapt it's code to the new exception structure.

Reviewer: ```suggestion
The Haskell community will have to adapt its code to the new exception structure.
```

As described in `Adding ``SomeExceptionWithBacktrace```_ the expected impact isn't
very high. This is further discussed in `Migration`_ .

Migration
---------

There was an intense discussion in the comments of the pull request of this
proposal about how to reach two goals
(<https://github.com/ghc-proposals/ghc-proposals/pull/330>):

> - Keep the migration costs as low as possible. I.e. most usages should work

Reviewer: ```suggestion
- Keep the migration costs as low as possible (i.e. most usages should work)
```

  without any change.
- Get a type checking error when the changed exception structure breaks existing

>   code. (I.e. no silent changes in behaviour.)

Reviewer: ```suggestion
  code (i.e. no silent changes in behaviour.)
```

The solution presented in this proposal has been agreed upon by all involved
parties.

``catch``and ``handle`` work with both, ``SomeExceptionWithBacktrace`` and
``SomeException``. This is the main reason for keeping ``SomeException``
as a layer in the exception hierarchy.

The usages of ``throw`` and ``throwIO`` don't have to be changed, too.
In fact, it showed that most submodules of GHC don't need any changes.
Only a few changes were needed to be made to GHC itself (e.g. in ``compiler/``).

Direct calls to ``toException`` and ``fromException``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As the types of ``toException`` and ``fromException`` changed, calls to them might
need to be adjusted to the new types. This is guided by type checking errors.

Usually, when these calls aren't bound to ``SomeException`` specific types, no
change should be needed.

In expressions where ``fromException``'s changed parameter type leads to type 
errors, one common solution is to convert the exception value first with
``toException``.

> As an example let's consider we don't want to change the type of 

Reviewer: ```suggestion
As an example let's consider the case of GHC's internal `GHC.TopHandler.real_handler` function; to ease compatibility, we may want to avoid changing the `SomeException` argument to `SomeExceptionWithBacktrace`. We can achieve this with a strategically-placed `fromException . toException`:
```

``GHC.TopHandler.real_handler`` to keep the needed migration effort small. ::

    real_handler :: (Int -> IO a) -> SomeException -> IO a
    real_handler exit se = do
      flushStdHandles -- before any error output
      -- The call to fromException needs to be preceded by a call to
      -- toException. (Original line from master commented out.)
      -- case fromException se of
      case (fromException . toException) se of
          Just StackOverflow -> do
              reportStackOverflow
              exit 2

          ... [Other cases]

Type synonym
~~~~~~~~~~~~

> As long as ``SomeExceptionWithBacktrace`` isn't deconstructed in pattern
> matches, a simple type synonym may be used to provide backwards
> compatibility. ::

Reviewer: ```suggestion
If no pattern matches are needed, compatibility with older compilers can be preserved by defining the `SomeExceptionWithBacktrace` type as a type synonym: ::
```

    module Control.Monad.Catch where

    #if __GLASGOW_HASKELL__ < 903
    type SomeExceptionWithBacktrace = SomeException
    #endif

    catchAll :: MonadCatch m => m a -> (SomeExceptionWithBacktrace -> m a) -> m a

    handleAll :: MonadCatch m => (SomeExceptionWithBacktrace -> m a) -> m a -> m a

This is only needed when ``SomeExceptionWithBacktrace`` should be used as type
in the program (e.g. to be able to access the backtraces).
As already discussed, in very most cases it is fine to stick with
``SomeException`` which is supported by old versions of GHC and those that
implement this proposal.

> In general, it should always be considered if a "down cast" to ``SomeException``

Reviewer: Yes, although it would be worthwhile pointing out that such a down-cast would lose backtrace information.

Author: @bgamari , sounds reasonable. Will do.

Author: Done.

using ``toException`` and ``fromException`` would not be sufficient to solve the
issue at hand.

Type and pattern synonym
~~~~~~~~~~~~~~~~~~~~~~~~

If there are existing pattern matches on the structure of ``SomeException``, a
combination of type and a pattern synonym could be applied. ::

    #if __GLASGOW_HASKELL__ < 903
    type SomeExceptionWithBacktrace = SomeException

    {-# COMPLETE SomeExceptionWithBacktrace #-}
    pattern SomeExceptionWithBacktrace :: forall. SomeException -> () -> SomeException
    pattern SomeExceptionWithBacktrace e unit <- (\x -> ((), x) -> (unit, e))
      where
        SomeExceptionWithBacktrace (SomeException e) _ = SomeException e
    #endif

> This should be considered the last resort! Please refer to `Alternatives`_

Reviewer: ```suggestion
This should be considered to be a measure of last resort! Please refer to `Alternatives`_
```
